test/hex_tarball_tests.erl

```diff
@@ -0,0 +1,205 @@
+
+    ExpectedRequirements = hex_tarball:normalize_requirements(Normal),
+    ExpectedRequirements = hex_tarball:normalize_requirements(Legacy),
+    ok.
+
+unpack_error_handling_test() ->
+    {ok, {Tarball, Checksum}} = hex_tarball:create(#{"name" => <<"foo">>}, [{"rebar.config", <<"">>}]),
+    {ok, #{checksum := Checksum}} = hex_tarball:unpack(Tarball, memory),
+    {ok, OuterFileList} = hex_erl_tar:extract({binary, Tarball}, [memory]),
+    OuterFiles = maps:from_list(OuterFileList),
+
+    %% tarball
+
+    {error, {tarball, eof}} = hex_tarball:unpack(<<"badtar">>, memory),
+
+    {error, {tarball, empty}} = unpack_files(#{}),
+
+    {error, {tarball, {missing_files, ["VERSION", "CHECKSUM"]}}} =
+        unpack_files(#{"metadata.config" => <<"">>, "contents.tar.gz" => <<"">>}),
+
+    {error, {tarball, {invalid_files, ["invalid.txt"]}}} =
+        unpack_files(#{"invalid.txt" => <<"">>}),
+
+    {error, {tarball, {bad_version, <<"1">>}}} =
+        unpack_files(OuterFiles#{"VERSION" => <<"1">>}),
+
+    {error, {tarball, invalid_checksum}} =
+        unpack_files(OuterFiles#{"CHECKSUM" => <<"bad">>}),
+
+    {error, {tarball, {checksum_mismatch, _, _}}} =
+        unpack_files(OuterFiles#{"contents.tar.gz" => <<"">>}),
+
+    %% metadata
+
+    Files1 = OuterFiles#{
+      "metadata.config" => <<"ok $">>,
+      "CHECKSUM" => <<"1BB37F9A91F9E4A3667A4527930187ACF6B9714C0DE7EADD55DC31BE5CFDD98C">>
+    },
+    {error, {metadata, {illegal, "$"}}} = unpack_files(Files1),
+
+    Files2 = OuterFiles#{
+      "metadata.config" => <<"ok[">>,
+      "CHECKSUM" => <<"0423D201115A49644F8BD4F216E666AF823CFE759853D9994CE9B652C5E604D9">>
+    },
+    {error, {metadata, invalid_terms}} = unpack_files(Files2),
+
+    Files3 = OuterFiles#{
+      "metadata.config" => <<"asdf.">>,
+      "CHECKSUM" => <<"F80D9B63D52695C6AC165D41F0F42F8D19152BE84D11FC2E5093FEC53CD4E3D9">>
+    },
+    {error, {metadata, {user, "illegal atom asdf"}}} = unpack_files(Files3),
+
+    Files4 = OuterFiles#{
+      "metadata.config" => <<"ok.">>,
+      "CHECKSUM" => <<"5E891D99F011F3DF8AB42E3B16420034C63CEC771B548F5E430057D13B62EF2B">>
+    },
+    {error, {metadata, not_key_value}} = unpack_files(Files4),
+
+    %% contents
+
+    Files5 = OuterFiles#{
+      "contents.tar.gz" => <<"badtar">>,
+      "CHECKSUM" => <<"9AA6A026CCD93D4E7B3B6083595259D2946709DE8F9EC7FAB69F2E53939DF403">>
+    },
+    {error,{inner_tarball,eof}} = unpack_files(Files5),
+
+    ok.
+
+%%====================================================================
+%% Helpers
+%%====================================================================
+
+in_tmp(F) ->
+    Old = file:get_cwd(),
+    TmpDir = "tmp",
+    ok = rm_rf(TmpDir),
+    ok = file:make_dir(TmpDir),
+    Dir = TmpDir ++ "/" ++ integer_to_list(erlang:unique_integer()),
+    ok = file:make_dir(Dir),
+    file:set_cwd(Dir),
+    apply(F, []),
+    file:set_cwd(Old).
+
+rm_rf(Path) ->
+    [] = os:cmd("rm -rf " ++ Path),
```

Comment:
this is ugly, need to find a cleaner way.

Reply:
Less important since it is a test, but in case anyone wants to contribute from Windows it'll need support. So something like https://github.com/erlang/rebar3/blob/master/src/rebar_file_utils.erl#L170